gcapi/retries.py

```diff
@@ -0,0 +1,64 @@
+from functools import wraps
+from typing import Dict, Optional
+
+import httpx
+from httpx import codes
+
+Seconds = float
+
+
+class BaseRetryStrategy:
+    def get_delay(self, latest_response: httpx.Response) -> Optional[Seconds]:
+        """
+        Returns the number of seconds to pause before the next retry,
+        based on the latest response.
+
+        Optionally, if None is returned no retry is to be performed.
+        """
+        raise NotImplementedError(
+            "The 'get_delay' method must be implemented."
+        )  # pragma: no cover
+
+
+NO_RETRY = None
+
+
+class SelectiveBackoffStrategy(BaseRetryStrategy):
+    """
+    Retries responses with codes of transient server errors (i.e. 5xx)
+    with an exponential backoff.
+
+    Each response code has its own backoff counter.
+    """
+
+    def __init__(self, backoff_factor, maximum_number_of_retries):
+        self.backoff_factor: float = backoff_factor
+        self.maximum_number_of_retries: int = maximum_number_of_retries
+        self.earlier_number_of_retries: Dict[int, int] = dict()
+
+    def __call__(self) -> BaseRetryStrategy:
+        return self.__class__(
+            backoff_factor=self.backoff_factor,
+            maximum_number_of_retries=self.maximum_number_of_retries,
+        )
+
+    @wraps(BaseRetryStrategy.get_delay)
```

Comment:
Since Python doesn't really do interfacing, the `@wrap` was just meant to copy over the docstring and annotations. Having played around a bit, it seems like the doc strings are actually pass through. Or at least `help()` returns the overwritten one.

Reply:
> Or at least help() returns the overwritten one.

Yes exactly! That was my point. They are not really "pass-through", it is a bit more complicated... the class-system uses a separate mechanism, which you actually you already reminded me of, with `mro`-stuff. :+1: